cpp/src/lists/segmented_sort.cu

```diff
@@ -0,0 +1,249 @@
+        auto const null_replace_T = null_precedence == null_order::AFTER
+                                      ? std::numeric_limits<T>::max()
+                                      : std::numeric_limits<T>::min();
+        auto device_child = column_device_view::create(child, stream);
+        auto keys_in =
+          cudf::detail::make_null_replacement_iterator<T>(*device_child, null_replace_T);
+        thrust::copy_n(rmm::exec_policy(stream), keys_in, child.size(), keys.begin());
+        return keys;
+      }
+      return rmm::device_uvector<T>{0, stream};
+    }();
+
+    std::unique_ptr<column> sorted_indices = cudf::make_numeric_column(
+      data_type(type_to_id<size_type>()), child.size(), mask_state::UNALLOCATED, stream, mr);
+    mutable_column_view mutable_indices_view = sorted_indices->mutable_view();
+    thrust::sequence(rmm::exec_policy(stream),
+                     mutable_indices_view.begin<size_type>(),
+                     mutable_indices_view.end<size_type>(),
+                     0);
+
+    if (column_order == order::ASCENDING)
+      SortPairsAscending(child.nullable() ? keys.data() : child.begin<T>(),
+                         mutable_output_view.begin<T>(),
+                         mutable_indices_view.begin<size_type>(),
+                         mutable_indices_view.begin<size_type>(),
+                         child.size(),
+                         offsets.size() - 1,
+                         offsets.begin<size_type>(),
+                         offsets.begin<size_type>() + 1,
+                         stream);
+    else
+      SortPairsDescending(child.nullable() ? keys.data() : child.begin<T>(),
+                          mutable_output_view.begin<T>(),
+                          mutable_indices_view.begin<size_type>(),
+                          mutable_indices_view.begin<size_type>(),
+                          child.size(),
+                          offsets.size() - 1,
+                          offsets.begin<size_type>(),
+                          offsets.begin<size_type>() + 1,
+                          stream);
+    std::vector<std::unique_ptr<column>> output_cols;
+    output_cols.push_back(std::move(output));
+    // rearrange the null_mask.
+    cudf::detail::gather_bitmask(cudf::table_view{{child}},
+                                 mutable_indices_view.begin<size_type>(),
+                                 output_cols,
+                                 cudf::detail::gather_bitmask_op::DONT_CHECK,
+                                 stream,
+                                 mr);
+    return std::move(output_cols.front());
+  }
+};
+
+std::unique_ptr<column> sort_lists(lists_column_view const& input,
+                                   order column_order,
+                                   null_order null_precedence,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  if (input.is_empty()) return {};
```

Comment:
This will return a null `unique_ptr`. Should use `empty_like`.

Reply:
updated.